OpenMarket/OpenMarket/Controller/ProductGridViewController.swift

```diff
@@ -0,0 +1,96 @@
+//
+//  ProductGridViewController.swift
+//  OpenMarket
+//
+//  Created by 예거 on 2022/01/12.
+//
+
+import UIKit
+
+class ProductGridViewController: UICollectionViewController {
+    
+    private var initialProductsListPage: ProductsListPage?
+    private let flowLayout = UICollectionViewFlowLayout()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureGridLayout()
+        
+        let request = ProductsListPageRequest(pageNo: 1, itemsPerPage: 20)
+        APIExecutor().execute(request) { (result: Result<ProductsListPage, Error>) in
+            switch result {
+            case .success(let productsListPage):
+                self.initialProductsListPage = productsListPage
+                DispatchQueue.main.async {
+                    self.collectionView.reloadData()
+                }
+            case .failure(let error):
+                // Alert 넣기
+                print("ProductsListPage 통신 중 에러 발생 : \(error)")
+                return
+            }
+        }
+    }
+    
+    private func configureGridLayout() {
+        collectionView.collectionViewLayout = flowLayout
+        let cellWidth = view.bounds.size.width / 2 - 10
+        flowLayout.itemSize = CGSize(width: cellWidth, height: cellWidth * 1.5)
+        flowLayout.minimumLineSpacing = 10
+        flowLayout.minimumInteritemSpacing = 10
+        flowLayout.scrollDirection = .vertical
+        flowLayout.sectionInset = UIEdgeInsets(top: 5, left: 5, bottom: .zero, right: 5)
+    }
+
+    // MARK: UICollectionViewDataSource
+
+    override func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 1
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return initialProductsListPage?.pages.count ?? .zero
+    }
+
+    override func collectionView(
+        _ collectionView: UICollectionView,
+        cellForItemAt indexPath: IndexPath
+    ) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(
+            withReuseIdentifier: ProductGridViewCell.reuseIdentifier,
+            for: indexPath
+        ) as? ProductGridViewCell else {
+            return ProductGridViewCell()
+        }
+        
+        guard let product = initialProductsListPage?.pages[indexPath.item] else {
```

Comment:
제가 담당했던 서비스들에서는 array indexing을 할 때 'index out of bounds' exception을 방지하기위해 subscript를 사용했습니다. 수정 필수는 아니고요, 참고만 해주세요 

https://stackoverflow.com/questions/25329186/safe-bounds-checked-array-lookup-in-swift-through-optional-bindings

Reply:
`index out of bounds` exception은 쉽게 일어날 수 있는 에러인 것 같아 별도 처리를 해주는 것이 안전하다!고 생각이 들었습니다.

저희가 실험을 해보았는데요👩🏻‍🔬
<img width="1398" alt="스크린샷 2022-01-17 오후 6 59 03" src="https://user-images.githubusercontent.com/81469717/149748717-ab342ac5-130d-4e69-a1e1-375ec5a9b47e.png">
위와 같이 range를 벗어난 index를 넣어주면 else문으로 빠지기도전에 Fatal Error가 발생하더라구요ㅠㅠ
그래서 참고주신 링크의 방법 처럼, range를 벗어난 index가 들어올 경우 `nil`이라도 반환시켜서 빈 cell을 반환하도록 수정해보겠습니다!

Reply:
배열(Array)에서 `index out of range`를 방지하기 위해, 찰리가 말씀해주신 `subscript(safe:)`를 Array 의 Extension 으로 구현했습니다!

링크에서는 조건문으로 삼항연산자를 사용하고 있는데요, 저희는 `switch문`을 사용해서, 좀 더 분명하게 로직을 구분해주었습니다.

```swift
extension Array {
    
    /// Returns the element at the specified index if it is within bounds, otherwise nil.
    subscript(safe index: Index) -> Element? {
        
        switch indices.contains(index) {
        case true:
            return self[index]
        case false:
            return nil
        }
    }
}

```